src/backend/env/mod.rs

```diff
@@ -1719,33 +1718,34 @@ impl State {
 
     fn charge_for_inactivity(&mut self, now: u64) {
         let mut inactive_users = 0;
-        let mut credits_total = 0;
         // Don't charge below this credit balance
         let inactive_user_balance_threshold = CONFIG.inactivity_penalty * 4;
-        for (id, credits) in self
-            .users
-            .values()
-            .filter(|user| !user.active_within_weeks(now, CONFIG.inactivity_duration_weeks))
-            .map(|u| (u.id, u.credits()))
-            .collect::<Vec<_>>()
-        {
+        let mut charges = Vec::new();
+        for user in self.users.values_mut() {
+            if !user.active_within_weeks(now, CONFIG.voting_power_activity_weeks) {
+                user.mode = Mode::Credits;
+            }
+            if user.active_within_weeks(now, CONFIG.inactivity_duration_weeks) {
+                continue;
+            }
             inactive_users += 1;
-            let costs = CONFIG
-                .inactivity_penalty
-                .min(credits.saturating_sub(inactive_user_balance_threshold));
+            let costs = CONFIG.inactivity_penalty.min(
+                user.credits()
+                    .saturating_sub(inactive_user_balance_threshold),
+            );
+            charges.push((user.id, costs));
+        }
+
+        let mut credits_total = 0;
+        for (user_id, costs) in charges {
             if costs > 0 {
-                if let Err(err) = self.charge(id, costs, "inactivity penalty".to_string()) {
+                if let Err(err) = self.charge(user_id, costs, "inactivity penalty".to_string()) {
                     self.logger
                         .warn(format!("Couldn't charge inactivity penalty: {:?}", err));
                 } else {
                     credits_total += costs;
                 }
             }
-            let user = self.users.get_mut(&id).expect("no user found");
-            user.change_rewards(
```

Comment:
Not needed anymore, because who're inactive longer than for `voting_power_activity_weeks` will be in credits mode and cannot have any rewards.